atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/expectations/TrackingKeyValueServiceReadInfoTest.java

```diff
@@ -0,0 +1,343 @@
+import com.palantir.atlasdb.transaction.api.expectations.ImmutableTransactionReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.TransactionReadInfo;
+import com.palantir.common.base.ClosableIterators;
+import com.palantir.util.paging.TokenBackedBasicResultsPage;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+/**
+ * {@link #size} must be divisible by 12 due to
+ * {@link TrackingKeyValueServiceTestUtils#createByteArrayByTableReferenceMapWithSize}.
+ * Refer to {@link TrackingKeyValueServiceTestUtils} javadoc.
+ * Also, the range of permitted sizes is constrained as some objects have size bounds.
+ * Note that the range of possible sizes will be larger if we use different sizes for different tests.
+ */
+@RunWith(Parameterized.class)
+public final class TrackingKeyValueServiceReadInfoTest {
+    @Parameterized.Parameters(name = "size = {0}")
+    public static Object[] size() {
+        return new Object[] {12 * 129, 12 * 365, 12 * 411};
+    }
+
+    private static final long TIMESTAMP = 14L;
+
+    @Rule
+    public MockitoRule rule = MockitoJUnit.rule();
+
+    @Mock
+    private List<byte[]> rows;
+
+    @Mock
+    private RangeRequest rangeRequest;
+
+    @Mock
+    private KeyValueService kvs;
+
+    private final int size;
+    private TrackingKeyValueService trackingKvs;
+    private final ImmutableMap<Cell, Value> valueByCellMapOfSize;
+
+    public TrackingKeyValueServiceReadInfoTest(int size) {
+        this.size = size;
+        this.valueByCellMapOfSize = TrackingKeyValueServiceTestUtils.createValueByCellMapWithSize(size);
+    }
+
+    @Before
+    public void setUp() {
+        this.trackingKvs = new TrackingKeyValueServiceImpl(kvs);
+    }
+
+    @Mock
+    private TableReference tableReference;
+
+    @Mock
+    private Map<Cell, Long> timestampByCellMap;
+
+    @Test
+    public void readInfoIsCorrectAfterGetAsyncCallAndFutureConsumption()
+            throws ExecutionException, InterruptedException {
+        when(kvs.getAsync(tableReference, timestampByCellMap))
+                .thenReturn(Futures.immediateFuture(valueByCellMapOfSize));
+
+        trackingKvs.getAsync(tableReference, timestampByCellMap).get();
+
+        validateReadInfoForReadForTable("getAsync");
+    }
+
+    @Test
+    public void throwingGetAsyncResultTracksNothing() {
+        when(kvs.getAsync(tableReference, timestampByCellMap))
+                .thenReturn(Futures.immediateFailedFuture(new RuntimeException()));
+
+        // ignore exception
+        try {
+            trackingKvs.getAsync(tableReference, timestampByCellMap).get();
+        } catch (Exception ignored) {
+        }
```

Comment:
perhaps a tidier way of doing this is
```java
assertThatThrownBy(() -> ...)
```

Reply:
happened to see this later on: note the semantics are slightly different in that as written you don't care whether trackingKvs.getAsync(...) threw an exception. Since this test does have that as an assumption of the setup, you *do* want to fail here.

The "legacy" way of writing this is 
```
try {
  bla();
  fail(); // IMPORTANT
} catch (Exception ignored) {
  // I'm OK
}
```

but, just use @Jolyon-S's recommended way, which is canonical across our repos: `assertThatThrownBy(() -> ...)`!